File: datasets/Part 9 - Dimensionality Reduction/Section 44 - Linear Discriminant Analysis (LDA)/lda_sbn.py
# -*- coding: utf-8 -*-
"""
Created on Wed Feb 19 15:35:38 2020

@author: jguerrero
"""

# LDA
# COMO IMPORTAR LAS LIBRERIAS
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# IMPORTAR EL DATASET
dataset = pd.read_csv('Wine.csv')

X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

# DIVIDIR EL DATASET EN ENTRENAMIENTO Y PRUEBA
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0) # random_state = 0 SOLO ES PARA QUE SIEMPRE SELECCIONE LOS MISMOS ELEMENTOS

# ESCALADO DE VARIABLES
# ESTANDARIZACION
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
# ESCALAR LOS DATOS DE TRAIN Y TEST CON LA MISMA FORMULA
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# REDUCCION DE DIMENSIONES CON LDA
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
lda = LDA(n_components = 2)
X_train = lda.fit_transform(X_train, y_train)
X_test = lda.transform(X_test)

# AJUSTAR LA REGRESION LOGISTICA CON EL CONJUNTO DE ENTRENAMIENTO
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

# PREDICCION DE LA REGRESION LOGISTICA CON EL CONJUNTO DE PRUEBA
y_pred = classifier.predict(X_test)

# EVALUAR RESULTADOS CON MATRIZ DE CONFUSION
from sklearn.metrics import confusion_matrix
conf_matrix = confusion_matrix(y_test, y_pred)

# VISUALIZACION DE LOS RESULTADOS
colors = ('red', 'green', 'blue')

from matplotlib.colors import ListedColormap
X_set, y_set = X_train, y_train
#X_set, y_set = X_test, y_test
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(colors))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(colors)(i), label = j)
plt.title('Clasificador (Conjunto de Entrenamiento)')
plt.xlabel('CP1')
plt.ylabel('CP2')
plt.legend()
plt.show()

File: datasets/Part 9 - Dimensionality Reduction/Section 44 - Linear Discriminant Analysis (LDA)/lda_sbn.R
# LDA
# Importar dataset
dataset <- read.csv('Wine.csv')


# Dividir el dataset en entrenamiento y prueba
#install.packages('caTools')
library(caTools)
set.seed(123) #SOLO ES PARA QUE SIEMPRE SELECCIONE LOS MISMOS ELEMENTOS
split <- sample.split(dataset$Customer_Segment, SplitRatio = 0.8)
trainning_set <- subset(dataset, split == TRUE)
testing_set <- subset(dataset, split == FALSE)

# Escalado de variables
trainning_set[, -14] <- scale(trainning_set[, -14])
testing_set[, -14] <- scale(testing_set[, -14])

library(MASS)

lda <- lda(formula = Customer_Segment ~ .,
           data = trainning_set)

trainning_set <- as.data.frame(predict(lda, trainning_set))
trainning_set <- trainning_set[, c(5, 6, 1)]

testing_set <- as.data.frame(predict(lda, testing_set))
testing_set <- testing_set[, c(5, 6, 1)]

library(e1071)
# Ajustar svm con el conjunto de entrenamiento
classifier <- svm(formula = class ~ .,
                  data = trainning_set,
                  type = "C-classification",
                  kernel = "linear")

# Prediccion de la regresion logistica con el conjunto de prueba
y_pred <- predict(classifier,
                  newdata = testing_set[, -3])

# Evaluar resultados con matriz de confusion
conf_matrix <- table(testing_set[, 3], y_pred)

# Visualizacion de los resultados
#install.packages('ElemStatLearn')
library(ElemStatLearn)
#set <- trainning_set
set <- testing_set
X1 <- seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.03)
X2 <- seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.03)
grid_set <- expand.grid(X1, X2)
colnames(grid_set) = c('x.LD1', 'x.LD2')
y_grid <- predict(classifier,
                  newdata = grid_set)
plot(set[, -3],
     main = 'Clasificador (Conjunto de Entrenamiento)',
     xlab = 'DL1',
     ylab = 'DL2',
     xlim = range(X1),
     ylim = range(X2))
contour(X1,
        X2,
        matrix(as.numeric(y_grid),
               length(X1),
               length(X2)),
        add = TRUE)
points(grid_set,
       pch = '.',
       col = ifelse(y_grid == 2, 'deepskyblue',
                    ifelse(y_grid == 1, 'springgreen3', 'tomato')))
points(set,
       pch = 21,
       bg = ifelse(set[, 3] == 2, 'blue3',
                   ifelse(set[, 3] == 1, 'green4', 'red3')))

#install.packages('ElemStatLearn')
library(ElemStatLearn)
set <- trainning_set
#set <- testing_set
X1 <- seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.02)
X2 <- seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.02)
grid_set <- expand.grid(X1, X2)
colnames(grid_set) = c('x.LD1', 'x.LD2')
y_grid <- predict(classifier,
                  newdata = grid_set)
plot(set[, -3],
     main = 'Clasificador (Conjunto de Prueba)',
     xlab = 'DL1',
     ylab = 'DL2',
     xlim = range(X1),
     ylim = range(X2))
contour(X1,
        X2,
        matrix(as.numeric(y_grid),
               length(X1),
               length(X2)),
        add = TRUE)
points(grid_set,
       pch = '.',
       col = ifelse(y_grid == 2, 'deepskyblue',
                    ifelse(y_grid == 1, 'springgreen3', 'tomato')))
points(set,
       pch = 21,
       bg = ifelse(set[, 3] == 2, 'blue3',
                   ifelse(set[, 3] == 1, 'green4', 'red3')))